First complete version reviewed

TheLokj opened the pull request.

TheLokj committed:
First commit

TheLokj committed:
Correction on the .sif path in README.md

TheLokj committed:
Merge branch 'main' into dev

TheLokj committed:
Improve the code structuration

TheLokj committed:
Improve the code structuration

TheLokj committed:
Merge branch 'dev' of https://github.com/TheLokj/kingdom-investigator into dev

TheLokj committed:
Adding the second step of validation using CAT

TheLokj committed:
Adding the second step of validation using CAT

TheLokj committed:
Update README.md

Vkale1 commented on README.md:
```diff
@@ -1,21 +1,25 @@
 
-In addition, although it is possible to specify the use of CUDA, please first check that the image used corresponds to the GPU version of DeepMicroClass. 
\ No newline at end of file
+In addition, although it is possible to specify the use of CUDA, please first check that the used image corresponds to the GPU version of DeepMicroClass.
```
An image of your pipeline + decisions it's making if you have one would be useful? - save it in the repo in a figures directory as well.

TheLokj replied:
Excellent idea ! I'll probably add a Mermaid diagram (DAG) and a schema summarizing the decision when we will have confirmed the precise stages of the pipeline :)

Vkale1 commented on bin/parse_dmc_classification.py:
```diff
@@ -0,0 +1,12 @@
+# -*- coding: utf-8 -*-
+import sys
+
+if " " in sys.argv[1] :
+  seqName = sys.argv[1].split(" ")[0]
+else :
+  seqName = sys.argv[1]
```
```suggestion
seq_name = sys.argv[1].split(" ", 1)[0]
```
Should hopefully capture the first instance before the space if it exists.

TheLokj replied:
Definitively better and easier yes, thanks !

Vkale1 commented on bin/parse_dmc_classification.py:
```diff
@@ -0,0 +1,12 @@
+else :
+  seqName = sys.argv[1]
+
+if float(sys.argv[2]) != 0 and float(sys.argv[3]) != 0 and float(sys.argv[4]) != 0 and float(sys.argv[5]) != 0 and float(sys.argv[6]) != 0 : 
+  print(f'{seqName}\t{["Eukaryote", "EukaryoteVirus", "Plasmid", "Prokaryote", "ProkaryoteVirus"][[float(sys.argv[2]), float(sys.argv[3]), float(sys.argv[4]), float(sys.argv[5]), float(sys.argv[6])].index(max(float(sys.argv[2]), float(sys.argv[3]), float(sys.argv[4]), float(sys.argv[5]), float(sys.argv[6])))]}')
```
```suggestion
# select DMC classification with maximum ratio
dmc_values = [float(arg) for arg in sys.argv[2:7]]
if 0 not in dmc_values:
  max_index = dmc_values.index(max(dmc_values))
  labels = ["Eukaryote", "EukaryoteVirus", "Plasmid", "Prokaryote", "ProkaryoteVirus"]
  print(f'{seq_name}\t{labels[max_index]}')
```
Suggestion to make it a little readable for someone new. Can I also check the logic here. Is it possible that one of the values could be 0 but it will still pass? In which case you just need to find the max without the first if statement?

TheLokj replied:
That's also definitively cleaner like that thanks.
I only saw 0.0 in the case where the contig wasn't classified (too many N, length<500pb) by the model, 0.0 for all classes. I think (but I will re-check) the score is everytime different from 0.0 when the model makes prediction, but I wrote this if statement to avoid an unfortunate statistics.

Vkale1 commented on bin/parse_dmc_classification.py:
```diff
@@ -0,0 +1,12 @@
+if float(sys.argv[2]) != 0 and float(sys.argv[3]) != 0 and float(sys.argv[4]) != 0 and float(sys.argv[5]) != 0 and float(sys.argv[6]) != 0 : 
+  print(f'{seqName}\t{["Eukaryote", "EukaryoteVirus", "Plasmid", "Prokaryote", "ProkaryoteVirus"][[float(sys.argv[2]), float(sys.argv[3]), float(sys.argv[4]), float(sys.argv[5]), float(sys.argv[6])].index(max(float(sys.argv[2]), float(sys.argv[3]), float(sys.argv[4]), float(sys.argv[5]), float(sys.argv[6])))]}')
+else :
+  print(f'{seqName}\tUnknown')
```
```suggestion
  print(f'{seq_name}\tUnknown')
```

Vkale1 commented on bin/save_metadata.py:
```diff
@@ -0,0 +1,32 @@
+                "Models" : { "Tiara" : 
+                                { "TiaraFirstStage" : sys.argv[6],
+                                "NeuralNetworkWeightsFirstStage" : sys.argv[7]},
+                            "DeepMicroClass" : sys.argv[8]}
```
Look great btw thanks for adding this! I will confirm.. but I think for the Models and DB files we should add an md5 as well.

Do I understand correctly that if we clone the DMC repo or update the tool with pip the models file will also update? I think we should have a static version in the databases directory to mitigate this since model path can be given as an argument.

TheLokj replied:
Yes that could be a great idea to have md5 ! And yes, the DeepMicroClass container use the model saved in the image, which is from the repository clone. The model is probably updated when the tool is then, so yes, keep a model outside the container can be a great idea to distinguish the both and avoid version problems.

Vkale1 commented on configs/codon.config:
```diff
@@ -0,0 +1,56 @@
+    diamondDB = "/hps/nobackup/rdf/metagenomics/service-team/ref-dbs/cat/CAT_prepare_20210107/Diamond_2.0.6/diamond"
+    taxonomyDB = "/hps/nobackup/rdf/metagenomics/service-team/ref-dbs/cat/CAT_prepare_20210107/2021-01-07_taxonomy/"
+    tiaraVersion = "1.0.3"
+    dmcVersion = "1.0.3"
```
Add the static models paths here as well once you have them (can be a temp dir in your hps location for now)

Vkale1 commented on modules/add_names_cat.nf:
```diff
@@ -0,0 +1,25 @@
+    script:
+    if (contig2classification =~ /cat.out.contig2classification.txt/) {
+    """
+    CAT add_names -i $contig2classification -o "cat.out.lineage" -t $taxonomyDB 
```
Does --exclude_scores remove the scores in the full lineage names? Since there is already a score field?

TheLokj replied:
Just tried and yes it does, but I don't know if it's really useful as we only use the superkingdom score (and as it's quicker to extract in the same time the superkingdom label)

TheLokj committed:
Update bin/parse_dmc_classification.py

Co-authored-by: Varsha Kale <34109092+Vkale1@users.noreply.github.com>

Vkale1 commented on modules/get_contigs_info.nf:
```diff
@@ -25,6 +25,6 @@ process GET_CONTIGS_INFO {
-    print header[1], l}' $contigsPath | tr -d ">" | sed "1d"
+    print header[1], l}' $contigsPath | tr -d ">" | sed "1d" > output.get_contigs_info
   """
 }
```
This is sequence ID and length?
Suggestion to use something existing e.g. seqtk just in case we end up with any weird scenarios in public contigs or long-read assemblies 
https://bioinf.shenwei.me/seqkit/usage/#fx2tab-tab2fx 
https://nf-co.re/modules/seqkit_fx2tab

TheLokj replied:
Yes this is it, I'll check to use seqkit then.

Vkale1 commented on modules/make_first_decision.nf:
```diff
@@ -0,0 +1,46 @@
+  path "output.make_first_decision" 
+  
+  script:
+  // Cases where DeepMicroClass and Tiara have the same prediction
+    // Classified as eukaryote
+  if (dmcPrediction == "Eukaryote" && tiaraPrediction == "eukarya") 
+    "echo -e '$id\t$length\t$tiaraPrediction\t$dmcPrediction\tClassifiedAsEukaryote' > output.make_first_decision"
+    // Classified as prokaryote
+  else if (dmcPrediction == "Prokaryote" && tiaraPrediction == "bacteria") 
+    "echo -e '$id\t$length\t$tiaraPrediction\t$dmcPrediction\tClassifiedAsProkaryote' > output.make_first_decision"
+  else if (dmcPrediction == "Prokaryote" && tiaraPrediction == "prokarya")
+    "echo -e '$id\t$length\t$tiaraPrediction\t$dmcPrediction\tClassifiedAsProkaryote' > output.make_first_decision"
+  else if (dmcPrediction == "Prokaryote" && tiaraPrediction == "archea") 
+    "echo -e '$id\t$length\t$tiaraPrediction\t$dmcPrediction\tClassifiedAsProkaryote' > output.make_first_decision"
+  else if (dmcPrediction == "Plasmid" && tiaraPrediction == "bacteria") 
+    "echo -e '$id\t$length\t$tiaraPrediction\t$dmcPrediction\tClassifiedAsProkaryote' > output.make_first_decision"
+  else if (dmcPrediction == "Plasmid" && tiaraPrediction == "prokarya")
+    "echo -e '$id\t$length\t$tiaraPrediction\t$dmcPrediction\tClassifiedAsProkaryote' > output.make_first_decision"
+  else if (dmcPrediction == "Plasmid" && tiaraPrediction == "archea") 
+    "echo -e '$id\t$length\t$tiaraPrediction\t$dmcPrediction\tClassifiedAsProkaryote' > output.make_first_decision"
+  
+  // Cases where Tiara classify as an eukaryote but DeepMicroClass don't
+  else if (dmcPrediction != "Eukaryote" && tiaraPrediction == "eukarya") 
+    "echo -e '$id\t$length\t$tiaraPrediction\t$dmcPrediction\tClassifiedAsEukaryote' > output.make_first_decision"
+
+  // Cases where DeepMicroClass classify as a prokaryote but Tiara don't
+  else if (dmcPrediction == "Eukaryote" && tiaraPrediction == "archea") 
+    "echo -e '$id\t$length\t$tiaraPrediction\t$dmcPrediction\tRequiringCATvalidation' > output.make_first_decision"
+  else if (dmcPrediction == "Eukaryote" && tiaraPrediction == "bacteria") 
+    "echo -e '$id\t$length\t$tiaraPrediction\t$dmcPrediction\tRequiringCATvalidation' > output.make_first_decision"
+  else if (dmcPrediction == "Eukaryote" && tiaraPrediction == "prokarya")
+    "echo -e '$id\t$length\t$tiaraPrediction\t$dmcPrediction\tRequiringCATvalidation' > output.make_first_decision"
```
```suggestion
dmc_prok = ["Prokaryote", "Plasmid"]
tiara_prok = ["prokarya", "archaea", "bacteria"]

  // Cases where DeepMicroClass and Tiara have the same prediction
    // Classified as eukaryote
  if (dmcPrediction == "Eukaryote" && tiaraPrediction == "eukarya") 
    "echo -e '$id\t$length\t$tiaraPrediction\t$dmcPrediction\tClassifiedAsEukaryote' > output.make_first_decision"
    
    // Classified as prokaryote
  else if (dmcPrediction in dmc_prok && tiaraPrediction in tiara_prok) 
    "echo -e '$id\t$length\t$tiaraPrediction\t$dmcPrediction\tClassifiedAsProkaryote' > output.make_first_decision"
  
  // Cases where Tiara classify as an eukaryote but DeepMicroClass don't
  else if (dmcPrediction != "Eukaryote" && tiaraPrediction == "eukarya") 
    "echo -e '$id\t$length\t$tiaraPrediction\t$dmcPrediction\tClassifiedAsEukaryote' > output.make_first_decision"

  // Cases where DeepMicroClass classify as a eukaryote but Tiara don't
  else if (dmcPrediction == "Eukaryote" && tiaraPrediction in tiara_prok) 
    "echo -e '$id\t$length\t$tiaraPrediction\t$dmcPrediction\tRequiringCATvalidation' > output.make_first_decision"
```
Double check the comment change is correct for the last one

Couple questions:
DMC euk/prok I assume also means viral for both?
Why is Tiara splitting prokarya from archaea and bacteria? Is there some confidence cutoff for the more specific predictions or does prokarya mean something else?
Are there potential cases where DMC tool predicts nothing?

Vkale1 replied:
I think the code could probably be simplified even further but please do check that I didn't mess up the logic. I haven't tested it.

TheLokj replied:
1. No it doesn't, that's why I was waiting for the viral classification test but according to the results I think it's better to ignore the virus scores and use the eukaryotes, prokaryotes and plasmids ones for this first step. 
2. The prokarya class is used when Tiara can't distinguish archea and bacteria. There is also the organella class, which can be ambiguous because of some organella origin. 
3. I indeed need to also add this case.

I'll write all the possibilities and add them to this script, always minimizing categorization as eukaryotic to avoid misclass of prokaryotic contigs. About the code itself, I don't know if it's possible as I never saw "in" used in Nextflow for the moment, but as I'm taking all these advices one by one I will can check it, thanks !

Vkale1 commented on configs/codon.config:
```diff
@@ -0,0 +1,56 @@
+    
+    // When updating the container, please be careful and update also the above related version parameters
+    withName: 'PREDICT_TIARA' {
+        container = '/hps/nobackup/rdf/metagenomics/service-team/users/louison/nextflow/project/bin/tiara1.0.3.sif'
+    }
+     withName: 'PREDICT_DEEPMICROCLASS' {
+        container = '/hps/nobackup/rdf/metagenomics/service-team/users/louison/nextflow/project/bin/deepmicroclass1.0.3.sif'
```
Do you have the Dockerfiles for these two containers if you made them yourself?

TheLokj replied:
I made only DeepMicroClass myself and yes I still have the Dockerfile (which is quite primitive).

Vkale1 commented on modules/parse_dmc_classification.nf:
```diff
@@ -1,25 +1,15 @@
 process PARSE_DMC_CLASSIFICATION {
   label 'light'
-  publishDir "${outdir}", mode: 'copy'
 
```
Have been given a suggestion to explicitly specify python containers here in any modules reqiuiring python... 
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/biopython:1.75':
        'quay.io/biocontainers/biopython:1.75' }"

TheLokj replied:
That's a nice suggestion thanks, I'll add it !

Vkale1 commented on modules/predict_deepmicroclass.nf:
```diff
@@ -15,7 +13,7 @@ process PREDICT_DEEPMICROCLASS {
+  path "${contigsPath.split("/")[-1]}_pred_${encoding}_${mode}.tsv", emit: predictionsPath
   path "dmc.log", emit: logPath
 
   script:
```
```suggestion
  script:
  params.command = "DeepMicroClass predict -i $contigsPath -o ./ -e $encoding -md $mode -d $device"

if (mode == "single") {
    params.command += " -sl $singleLen"
}

if (modelPath != "default") {
    params.command += " -m $modelPath"
}

"""
echo "Params : $params.command" > dmc.log
$params.command
"""

```

I'm not sure I understand what mode single v hybrid is?

Vkale1 commented on modules/save_contigs_per_kingdom.nf:
```diff
@@ -3,15 +3,16 @@ process SAVE_CONTIGS_PER_KINGDOM {
 
   """
-  grep -Pzo '>${result[0]}\\s[^\\>]*' $contigsPath | sed '\$ s/.\$//'
+  grep -Pzo '>${result[0]}\\s[^\\>]*' $contigsPath | sed '\$ s/.\$//' > output.save_contigs_per_kingdom
```
If I understand this correctly you're saving fasta's per first prediction. Since you have the TSV with all the prediction information I think it is enough to have:
- an intermediate fasta for CAT confirmiation (not a final output file)
- two final output files eukaryotes.fasta other_kingdoms.fasta or something similar.

TheLokj replied:
I'm not sure this is what you're talking about, but the current script already save fastas during the first step as firstStep_contigs[firstPredictionClass].fa and, for non-sure prediction, as firstStep_RequiringCATvalidation.fa, and then after the second step as secondStep_Eukaryotes.fa and secondStep_NotEukaryotes.fa. 

So if I've understood your message correctly, that's already the case ! :)
I just need to change the fasta names, which aren't very pretty.

Vkale1 commented on modules/save_metadata.nf:
```diff
@@ -0,0 +1,23 @@
+
+    script:
+    """
+    python3 $projectDir/bin/save_metadata.py $tiaraVersion $dmcVersion \$(grep "# CAT" $logCat | cut -d"v" -f2 | sed 's/.\$//') \$(grep "Prodigal V" $logCat | cut -d"V" -f2 | cut -d":" -f1) \$(grep "diamond version" $logCat | cut -d" " -f7 | sed 's/.\$//') $catDB $diamondDB $taxonomyDB \$(grep "\\/" $logTiara | cut -f3 | sed -n 1p) \$(grep "\\/" $logTiara | cut -f3 | sed -n 3p) \$(grep "1/3" $log ...
```
If you're able to set the python script to take paths/variables as arguments this would be cleaner and future proof in case we add new tools. Then you can also remove the 'v' within the script just once.

TheLokj replied:
Yes that's dirty I agree ! Do you mean by using _argparse_ with the bash command like currently or by getting directly the information in the Python script with _subprocesses_ running the bash command (or even just by getting the information by Python parsing) ?

Vkale1 reviewed (changes requested):
Looks good thank you! 
Have added some comments about the pipeline. I think @vagkaratzas might have more useful comments regarding the Nextflow. We can have a chat about the format of the metadata json in person - couple of tweaks there.

vagkaratzas reviewed (changes requested):
Add .gitignore file with these as bare minimum:
        .nextflow*
        out/
        work/
        
See other pipelines for example: https://github.com/EBI-Metagenomics/mettannotator/blob/main/.gitignore

vagkaratzas commented on modules/add_names_cat.nf:
```diff
@@ -0,0 +1,25 @@
+process ADD_NAMES_CAT {
+    label 'light'
+    publishDir "${outdir}/CAT", mode: 'copy'
+    container = "/hps/nobackup/rdf/metagenomics/service-team/singularity-cache/quay.io_microbiome-informatics_cat:5.2.3.sif"
```
Remove container from here; already with the rest in the config file

vagkaratzas reviewed (commented):
Good stuff. Can you also add on the READme the command to run the simple test?

TheLokj committed:
Replace sed and awk script by SeqKit in get_contigs_info.nf and save_contigs_per_kingdom.nf

Update README.md and add the decision rule table

Change decision rules

Add Python containers for Pythin script

Update the metadata.json structure and make the save_metadata script more human-readable

Remove the duplicate CAT container

Simplify some scripts

TheLokj committed:
Merge branch 'dev' of https://github.com/TheLokj/kingdom-investigator into dev

TheLokj committed:
Remove conflicts

TheLokj committed:
Upload the current DAG

TheLokj committed:
Add the work in progress DAG in the README.md

TheLokj committed:
parse_dmc_classification.py correction

TheLokj committed:
Optimize the pipeline by launching only one time the decisions and save_contigs modules {launched for each contig previously)

Rename step to stage and change the fasta names to lower case ones

TheLokj committed:
Add the files related to the previous commit

TheLokj committed:
Change seqkit from -p to -f to avoid too many args limit

TheLokj committed:
Replace process ressource labels with the nf-core naming convention

TheLokj committed:
Replace process ressource labels with the nf-core naming convention

TheLokj committed:
Update CAT ressource allocation to make it compatible with real use case and avoid a OOM kill
